### PR TITLE
ProgressTableLevelBubble caching

### PR DESCRIPTION
**Pull request by cforkish**
while investigating the bug fixed by #40177 i started by exploring if it was a rendering or performance issue. that issue ended up being unrelated to rendering/performance, but as part of my exploration i took a quick pass at a performance optimization i've been considering for awhile: caching the frequently-reused parts of the bubbles.

aside from the url the bubbles link to, the vast majority of them differ only in a small set of unique content/styling options. for example, each row of the table displays bubbles with identical content as the other rows, and often even identical to other cells. and considering that many bubbles are often in the `not_tried` state, huge portions of the table are often duplicated.

this PR introduces caching the raw HTML for each unique bubble, and rendering that directly. this introduces a lot of additional complexity to the component, but also results in a noticeable improvement in scrolling performance.

my thinking is that if we like this approach, i will move all the caching logic into the `BubbleFactory` concept we've been mentioning, and use it only in the progress table. i'm happy to hop on a call to discuss the pros and cons if this approach if desired.

### Note:
while cleaning up this PR i discovered that our original implementation of this component broke the hover effect, because Radium doesn't work with functional components. this PR solves that by replacing the Radium styling with a new CSS class, since Radium also can't do anything with the raw HTML. if we decide not to move forward with this PR i will open a new PR to specifically address the hover issue.

### CODE REVIEW

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (maureensturgeon) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: Did you consider taking it one step further and caching by bubble shape and level number instead of `levelId`? I think this is a good starting point, I'm just curious.

**Review by maureensturgeon — COMMENTED** [review]
I'm for adding this change! I'll wait for James to take a look though before I approve. I'm curious, were you able to measure the performance impact?

**Comment by cforkish** [comment]
thanks to @maureensturgeon's suggestion, i moved away from level id as the primary cache key, instead focusing on bubble content, which is frequently repeated across lessons while the level ids are not. (note that my comment "and often even identical to other cells" in the PR description didn't actually apply while i was using level id).

this change reduced the size of the cache in the "small section, large script" storybook (which has 30 students) by an order of magnitude (~1500 -> ~150), and resulted in a very noticeable improvement in scrolling performance.

**Comment by cforkish** [comment]
i have caught a number of cache-related bugs during my testing, so if we decide to move forward with this i want to spend some time adding more tests, but i will wait for @jamescodeorg to review before doing that work.

i will just say though that after applying Maureen's suggestion, i _really_ think the performance improvement is worth the complexity. but i would definitely move it into a table-specific cacheable component in a subsequent PR so it doesn't muddy the general use cases.

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (cforkish) [reply apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: 🎉

**Comment by cforkish** [comment]
@maureensturgeon re. measuring performance, i tried, but was largely unsuccessful. what i was seeing made me think that the number of tabs i had open was impacting the results, which made it hard to compare. to do it right i'd need to figure out the best way to do a repeatable test that isn't impacted by environment. if you have any ideas for that i'm happy to discuss!

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: Live dangerously.

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: Mostly curious, why can't we use JSX here?

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: Why do we need to include levelId in the cache key?  (Does levelId ever affect how the bubble looks?)

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx:127]: I'm a little worried about maintaining this over time.  How effective would the cache be if we just compared all the props except for levelId and url?  (I'm wondering if we can just declare props to be in the cache key or out of the cache key as a way to simplify the design.)

**Review by jamescodeorg — COMMENTED** [review]
I like the idea.  I'm wondering if there's built-in functionality or a library that we can use to help manage the caching.  I saw that there's a `React.memo` but didn't get a chance to see if that would meet our needs here.  If there's nothing off-the-shelf, it might be worthwhile to implement this in a way that is reusable.  For example, cacheable components can implement a getCacheKey() method and an HoC or something similar would manage the caching.

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (cforkish) [reply apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: 🤘😝🤘

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (cforkish) [reply apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: thanks for asking, turns out we can! i took me a few tries to get the syntax right for getting the raw HTML, and at one point i thought i concluded this was necessary. but apparently it's not, so i will update to make this code ever so slightly less obtuse

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (cforkish) [reply apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: at this point that is just a fallback. my original implementation was only using `levelId` and `status` for the key, because the only difference between the appearance of the same level would come from status (actually, we also need `isPaired`, which i'd missed in my original implementation). 

i didn't want key generation and lookup to become a performance hit when we're rendering thousands of these at once. but Maureen pointed out that that resulted in lots of duplicated bubbles in the cache (think if the number of not tried bubbles with a number in them repeated across lessons). switching from id + status to this implementation shrunk the average cache size in storybook from ~1500 to ~150.

i left `levelId` as a fallback since i'd already added it in all the places where this code is called, but that was lazy of me. if we still aren't using it once we settle on our cache key implementation, i will remove it before merging.

personally, i think keeping the keys as small as possible is worth the potential extra maintenance cost. _all_ of the slowness in this table comes from js scripting work, so imo every little bit of optimization here is worth it. but if you'd prefer a more future-proof implementation i can revisit and try to measure the performance impact.

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (cforkish) [reply apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx:127]: as per above, that was my original implementation.... but there are _a lot_ of necessary props if we go that route. happy to discuss.

**Comment by cforkish** [comment]
@jamescodeorg re. your high-level feedback -- React.memo isn't supported in our version of React, but also i don't think it would be quite as performant as this. from my understanding it's more of a replacement for `React.PureComponent` when using functional components, which means it's more about avoiding unnecessary re-renders than reusing "rendered" html. once we are able to use it i think it would be worth investigating how it performs compared to this implementation.

re. libraries -- imo it's not worth looking into adding any new libraries until we upgrade react. once we do that, we can update our table library, which might make this level of optimization unnecessary in the first place (no idea what fancy performance optimizations are available to developers of modern react libraries, but i'm optimistic). so i'd vote for tabling that conversation for now.

re. abstractions -- i can't think of any other places in our codebase where something like this would be useful, so i'm not sure it's worth abstracting until we encounter another use case. i did think about it though, and am not opposed to the idea, but not sure it fits into this SOW. could be something to add to the list of requests for the component library!

for both of those ideas, something i'd be curious to explore at some point down the road would be generating SVGs to cache instead of HTML.

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (cforkish) [reply apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx:127]: documenting our offline conversation, i tested several implementations that use all props for easier maintainability and abstraction, but found that this super-custom key generation is noticeably more performant, and imo worth the additional complexity

**Comment by cforkish** [comment]
@jamescodeorg @maureensturgeon a few notes about the latest commit:

to demonstrate how we can keep the caching logic out of our main progress bubble code paths, i went ahead and took the first step toward building a `BubbleFactory`, which will be used to build `BasicBubble` components for different use cases. now that most of the work being done in `ProgressTableLevelBubble` is related to caching, which is functionality specific to the progress table bubbles, i intend to leave that component more-or-less as is as i continue updating our other progress bubbles to leverage `BubbleFactory`. as `BubbleFactory` evolves i may be able to move even more code out of `ProgressTableLevelBubble`, but my main point is that from here on out, `ProgressTableLevelBubble` is a component specifically designed for the progress tables, but sharing as much code as possible with other bubble components via `BubbleFactory` (rather than the previous plan, which was to eventually merge `ProgressTableLevelBubble` and `ProgressBubble` since we expected the table component to require a _subset_ of bubble functionality, whereas now it encompasses a _superset_).

i still need to update and add tests, but wanted to get eyes on this before i do all of that in case there are other changes or reversions we decide to make.

File: apps/src/util/CachedElement.jsx
Comment (jamescodeorg) [inline apps/src/util/CachedElement.jsx:14]: Should we put a comment here about why we think this is not dangerous?

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: (minor) Maybe update these two lines too?

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: typo: pr --> or

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: (minor) I wonder if it would be simpler to have the element that's cached just be a children of `<CachedElement>`.  Up to you.

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: Mostly curious -- did you put isBubbleEnabled in BubbleFactory because you expect it to be called from other places?

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: (very minor) Do you think moving this note to just above propTypes would make it more likely that it's noticed when a prop is added/removed?

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: Let's pick a different abbreviation for assessment.

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx:111]: Somewhere in this header comment, let's indicate that the "requirements" for the cache key / how to determine if a property should be included in the cache key.

File: apps/src/util/CachedElement.jsx
Comment (jamescodeorg) [inline apps/src/util/CachedElement.jsx:1]: This component ended up being fairly simple but I like how it separates out this functionality.

File: apps/src/templates/progress/BubbleFactory.jsx
Comment (jamescodeorg) [inline apps/src/templates/progress/BubbleFactory.jsx:1]: I think the functionality here ended up being less encapsulated than I imagined in my head -- this feels more like a set of helper functions than a factory pattern that abstracts the creation of bubble components.  I don't have a clear picture of how the other bubbles are the same/different as the ones in progress table so I'm curious to see how this evolves as we bring in the other bubbles.

**Review by jamescodeorg — COMMENTED** [review]
General direction looks good!

File: apps/src/templates/progress/BubbleFactory.jsx
Comment (cforkish) [reply apps/src/templates/progress/BubbleFactory.jsx:1]: i agree. this was just a quick pass at moving common bubble stuff out of `ProgressTableLevelBubble`, which had accumulated a decent number of internal helpers. so much of our bubble code is devoted to translating props (e.g. bubble should be disabled if `isDisabled || status === locked`), so i may end up moving `BasicBubble` into its own module and replacing `BubbleFactory` with `bubbleHelpers` or something like that. i might like to schedule a call with you and Maureen to discuss options as i develop a clearer picture of the varying needs between bubble use cases.

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (cforkish) [reply apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: 👍

Comment (cforkish) [reply apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: 👍

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (cforkish) [reply apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: since the children get passed into the parent as a prop, i was assuming that doing it that way would result in a call to `createElement` for the child even when we have it in the cache, which is what this component is intended to avoid. i'm sure that digging in to confirm that behavior would result in some valuable learnings, but i'm going to stick with this for now in the interest of time.

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (cforkish) [reply apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: yes, this translation from two props to one will be a common requirement.

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (cforkish) [reply apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx:111]: i was having trouble coming up with a succinct explanation of those requirements, so i was hoping that all the inline comments i added would provide "explanation by example". but i should at least add a note about the requirements relating to unique visual representation, which so far isn't mentioned anywhere.

File: apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
Comment (cforkish) [reply apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx]: yeah

File: apps/src/util/CachedElement.jsx
Comment (cforkish) [reply apps/src/util/CachedElement.jsx:1]: me too. it's also easy to imagine how this could be updated to provide a default implementation of key generation in the future.

File: apps/src/util/CachedElement.jsx
Comment (cforkish) [reply apps/src/util/CachedElement.jsx:14]: sgtm

**Review by cforkish — COMMENTED** [review]
thanks @jamescodeorg, great comments. will let you know once i've made these updates.

File: apps/test/unit/templates/sectionProgress/ProgressTableLevelBubbleTest.jsx
Comment (jamescodeorg) [inline apps/test/unit/templates/sectionProgress/ProgressTableLevelBubbleTest.jsx:75]: What does this line do?

File: apps/test/unit/util/CachedElementTest.jsx
Comment (jamescodeorg) [inline apps/test/unit/util/CachedElementTest.jsx]: (minor) Not sure if it makes a difference, but I like doing resets in beforeEach() to ensure that things in a particular state for the current test (as opposed to assuming that the previous test cleaned up properly).

(same comment for other test file)

**Review by jamescodeorg — APPROVED** [review]

File: apps/test/unit/util/CachedElementTest.jsx
Comment (cforkish) [reply apps/test/unit/util/CachedElementTest.jsx]: i went back and forth on this. it's also important that we clean up after ourselves, so i'll update this to `beforeEach` and add an `after` (called after this whole block of tests) that does the same.

File: apps/test/unit/templates/sectionProgress/ProgressTableLevelBubbleTest.jsx
Comment (cforkish) [reply apps/test/unit/templates/sectionProgress/ProgressTableLevelBubbleTest.jsx:75]: renders the bubble we want to test, which internally calls `createBubbleElement` so we can get the return value. i'll add a comment

Comment (jamescodeorg) [reply apps/test/unit/templates/sectionProgress/ProgressTableLevelBubbleTest.jsx:75]: Oh, I see.  I missed that the places that are calling this method are no longer mounting the component-under-test themselves anymore.

File: apps/src/templates/sectionProgress/progressTables/ProgressTableDetailCell.story.jsx
Comment (maureensturgeon) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableDetailCell.story.jsx:145]: Why was this story removed?

File: apps/test/unit/templates/sectionProgress/ProgressTableLevelBubbleTest.jsx
Comment (maureensturgeon) [inline apps/test/unit/templates/sectionProgress/ProgressTableLevelBubbleTest.jsx:83]: Typo `verifty`

**Review by maureensturgeon — APPROVED** [review]

**Pull request merged**
